src/rsh/mod.rs

```diff
@@ -59,51 +59,147 @@ pub fn run(initial_state: State) {
     }
 }
 
+#[derive(PartialEq)]
+enum BuildType {
+    None,
+    Single,
+    Double,
+}
+
+struct ParseResult {
+    result: Vec<String>,
+    completed: bool,
+    build_string: String,
+    build_type: BuildType,
+}
+
 fn parse_args(args: &String) -> Vec<String> {
-    let mut result: Vec<String> = Vec::new();
+    let mut parse_result: ParseResult = ParseResult {
```

Comment:
Should move all of this below line 88 and on the early return if `args.len() == 0` just return `Vec::new()` this will avoid all of these extra allocations.